platforms/unix/vm-display-fbdev/sqUnixEvdevKeyMouse.c

```diff
@@ -106,13 +110,39 @@ static void setSqueakModifierState();
 
 /* Mouse */
 
+#define NUM_SUPPORTED_SLOTS 32 /* We need to use EVIOCGABS to discover the actual number of
+                                  slots on a device, so this hardcoded limit is lazy */
 struct ms
 {
   char 		  *msName;
   int		   fd;
   struct libevdev *dev;
+
+  /*
+    Emulation of a relative mouse device using a (multi- or single-touch!) absolute device.
+    See Multitouch.md.
+  */
+
+  double touch_change_time; /* used for tap detection and drag lock */
+  int touching;         /* tracks BTN_TOUCH; we use this in tap detection. */
+  int moved;
+  int old_x;            /* -1 for none */
+  int old_y;            /* -1 for none */
+  int new_x;
+  int new_y;
+
+  int abs_type;         /* 0 => simple (singletouch), 1 => type A, 2 => type B */
+  int slots[NUM_SUPPORTED_SLOTS];
+  int primary_tracking_id; /* -1 for none */
+  int current_slot;     /* -1 for none */
+
+  double tap_release_deadline;
+  int tap_bit;
 };
 
+#define TAP_START_TIMEOUT 0.18
```

Comment:
Would it make sense to deliver the raw touch events from the plugin to the image and interpret them there as mouse events? That way the programmer in the image would have maximum control and you would not need to hardcode constants in the VM. Sorry if my question does not make sense, I have only briefly skimmed the diff.

Reply:
Absolutely it would, but that would be a much more invasive change to the VM infrastructure than I have time for right now. On the other hand, I have already had good results from doing all this directly within the image: see package LinuxIO at https://squeaksource.com/SqueakPhone/, which does finger tracking etc by reading `/dev/input/event*` devices from within the image.

Reply:
The hardcoding of constants is something to address though. Suggestions welcome (yet more command-line arguments??).